include/cuco/detail/static_multimap/device_view_impl.inl

```diff
@@ -1055,6 +1055,187 @@ class static_multimap<Key, Value, Scope, Allocator, ProbeSequence>::device_view_
     }  // while running
   }
 
+  /**
+   * @brief Retrieves all the matches of a given pair contained in multimap using vector
+   * loads without shared memory buffer.
+   *
+   * For pair `p`, if pair_equal(p, slot[j]) returns true, copies `p` to unspecified locations
+   * in `[probe_output_begin, probe_output_end)` and copies slot[j] to unspecified locations in
+   * `[contained_output_begin, contained_output_end)`. If `p` does not have any matches, copies
+   * `p` and a pair of `empty_key_sentinel` and `empty_value_sentinel` into the output only if
+   * `is_outer` is true.
+   *
+   * @tparam is_outer Boolean flag indicating whether outer join is peformed
+   * @tparam uses_vector_load Boolean flag indicating whether vector loads are used
+   * @tparam ProbingCG Type of Cooperative Group used to retrieve
+   * @tparam OutputIt1 Device accessible output iterator whose `value_type` is constructible from
+   * `InputIt`s `value_type`.
+   * @tparam OutputIt2 Device accessible output iterator whose `value_type` is constructible from
+   * the map's `value_type`.
+   * @tparam PairEqual Binary callable type
+   * @param probing_cg The Cooperative Group used to retrieve
+   * @param pair The pair to search for
+   * @param probe_output_begin Beginning of the output sequence of the matched probe pairs
+   * @param contained_output_begin Beginning of the output sequence of the matched contained
+   * pairs
+   * @param pair_equal The binary callable used to compare two pairs for equality
+   */
+  template <bool is_outer,
+            bool uses_vector_load,
+            typename ProbingCG,
+            typename OutputIt1,
+            typename OutputIt2,
+            typename PairEqual>
+  __device__ __forceinline__ std::enable_if_t<uses_vector_load, void> pair_retrieve(
+    ProbingCG const& probing_cg,
+    value_type const& pair,
+    OutputIt1 probe_output_begin,
+    OutputIt2 contained_output_begin,
+    PairEqual pair_equal) noexcept
+  {
+    using cuda::std::memory_order_relaxed;
+
+    auto const lane_id                = probing_cg.thread_rank();
+    auto current_slot                 = initial_slot(probing_cg, pair.first);
+    [[maybe_unused]] auto found_match = false;
+
+    __shared__ atomic_ctr_type counter;
```

Comment:
I think we can avoid needing a shared memory counter here by using `ballot` on `first_equals/second_equals` and can compute how many matches the other threads have to determine where to write this threads contribution.

Reply:
Right, we did something similar in the very early implementation of multimap.